packages/@react-spectrum/textfield/src/TextArea.tsx

```diff
@@ -29,28 +30,101 @@ function TextArea(props: SpectrumTextFieldProps, ref: RefObject<TextFieldRef>) {
     ...otherProps
   } = props;
 
+  let {labelPosition} = otherProps;
+
+  let fallbackRef = useRef<TextFieldRef>(null);
+  if (!ref) {
+    ref = fallbackRef;
+  }
+
   // not in stately because this is so we know when to re-measure, which is a spectrum design
   let [inputValue, setInputValue] = useControlledState(props.value, props.defaultValue, () => {});
 
   let inputRef = useRef<HTMLTextAreaElement>();
 
+  // Store styleProps to apply styles appropriately on height change.
+  let {styleProps} = useStyleProps(props);
+  let initialHeightsRef = useRef(null);
+
   let onHeightChange = useCallback(() => {
+    let input = inputRef.current;
+    let inputContainer = input.parentElement;
+    let field = ref.current.UNSAFE_getDOMNode();
+
+    // current offsetHeight of the textarea
+    let inputOffsetHeight = input.offsetHeight;
+
+    // Label or helpText height might change, so we have to recalculate
+    let labelHeight = labelPosition !== 'side' && !field.firstElementChild.contains(input) ? (field.firstElementChild as HTMLElement).offsetHeight : 0;
+
+    // With labelPosition === 'side', the TextArea and HelpText will be contained within a wrapper.
+    let fieldWrapper = labelPosition === 'side' ? field.lastElementChild : field;
+    let helpTextHeight = !fieldWrapper.lastElementChild.contains(input) ? (fieldWrapper.lastElementChild as HTMLElement).offsetHeight : 0;
+
+    // store a ref containing the initial size values for comparison
+    if (initialHeightsRef.current === null) {
+      initialHeightsRef.current = {
+        inputHeight: inputOffsetHeight,
+        fieldHeight: field.offsetHeight
+      };
+    }
+    let {inputHeight, fieldHeight} = initialHeightsRef.current;
+    let calculatedFieldHeight = Math.max(inputHeight, inputOffsetHeight) + labelHeight + helpTextHeight;
+    let heightStyles = Object.entries(
+      ['height', 'maxHeight', 'minHeight']
+      .reduce((obj, key) => {
+        let value = styleProps.style[key];
+        if (value && value !== '') {
+          obj[key] = value;
+        }
+        return obj;
+      }, {})
+    );
     if (isQuiet) {
-      let input = inputRef.current;
       let prevAlignment = input.style.alignSelf;
       input.style.alignSelf = 'start';
       input.style.height = 'auto';
-      input.style.height = `${input.scrollHeight}px`;
+      input.style.height = `${Math.max(input.scrollHeight, inputHeight)}px`;
       input.style.alignSelf = prevAlignment;
+      inputContainer.style.flexGrow = '0';
+      for (const [key, value] of heightStyles) {
+        switch (key as string) {
+          case 'height':
+          case 'maxHeight':
+            field.style[key as string] = calculatedFieldHeight <= fieldHeight ? value : '';
+            break;
+        }
+      }
+    } else {
```

Comment:
@snowystinger Adding a `helpTextHeightRef` to keep track of the helpText height on resize and adding the following here fixes the wiggling issue when the textarea has been manually resized:
```ts
      // If the textarea has been explicitly resized by the user,
      // and the helpText has resized to a new height after validation, 
      // adjust the input.style.height by the change in the helpText height.
      if (helpTextHeightRef.current !== helpTextHeight) {
        if (input.style.height) {
          inputOffsetHeight = parseFloat(input.style.height) - (helpTextHeight - helpTextHeightRef.current);
          input.style.height = `${inputOffsetHeight}px`;
        }
        helpTextHeightRef.current = helpTextHeight;
      }
```